Improve NameFilters

eatkins opened the pull request:
I'm working on adding proper scoping to WatchSources and this may
require de-duplication of redundant watch sources. To facilitate this,
I'm improving the toString/equals/hashCode methods of a number of
filters so that reporting is better and so that we can discard duplicate
sources (which will have to be compared using their include and exclude
filters).

eatkins committed:
Overhaul the file filters

The io filters as previously written do not always nicely override the
toString method and they also generally do not override equals and
hashCode. This makes it difficult to verify if any of the sources
generated by sbt are duplicates.

To address this, I introduce three new (sealed) traits:
1) AbstractAndFilter
2) AbstractNotFilter
3) AbstractOrFilter

These three traits represent the result of combining two filters or
inverting a single filter. They automatically override the equals,
hashCode and toString methods to sensible values. Each trait has a final
class that extends it for both regular FileFilters and NameFilters.

I also added equals and hashCode methods to Simple(File?)Filter.

typesafe-tools commented:
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=develop&zinc=develop&io=pull/204/head&librarymanagement=develop&util=develop&website=develop&deploy_to_bintray=false&run_sbt_community=false&verbose_dbuild=false&community_verbose=false">this link.</A>

eatkins committed:
Override Source.{ equals, hashCode }

I will be using this to de-duplicate watchSources in sbt.

eatkins commented:
I have some other filter improvements that I'm still working on. I'll remove WIP form the PR title when it's ready to go.

eatkins committed:
Add extension filter

It is very common in sbt to want to filter files with a particular
extension, or files whose extension is in some set of extensions. I
introduce the ExtensionFilter to special case this. I don't really
expect users to directly make an ExtensionFilter, but now when a user
creates a filter like:
watchSources in Compile += Source(sourceDirectory.value, "*.scala" | "*.java", NothingFilter)
the Source will actually be create with an ExtensionFilter. The
advantage of this is that regex matching is slow, whereas finding the
extension of a file is very fast. Once we have the extension, we can
just check if it is in the set of accepted extensions.

eatkins commented:
This is good to go now. After this change, all of the default watch sources in sbt are de-duplicatable and have nicer toString output.
